### ISSUE TITLE
Incorrect gradients for F.interpolate on GPU when mode=nearest

### ISSUE THREAD

**Issue by CedricPicron** [pytorch/pytorch#51036]
## 🐛 Bug

Function `F.interpolate` gives incorrect gradients on GPU when using `mode='nearest'`.

## To Reproduce

``` python
import torch
import torch.nn.functional as F


for seed in range(0, 20):
    torch.manual_seed(seed)

    in_shape = torch.randint(low=10, high=50, size=(2,)).tolist()
    in_map = torch.randn(2, 256, *in_shape, device='cuda', requires_grad=True)

    out_shape = torch.randint(low=10, high=50, size=(2,)).tolist()
    inter_map = F.interpolate(in_map, size=tuple(out_shape), mode='nearest')

    output_map = torch.mul(inter_map, 0.0)
    output_map.sum().backward()

    max_grad = in_map.grad.abs().max().item()
    print(f"({seed:02d}) Max grad: {max_grad}")
```

## Output on my GPU

```
(00) Max grad: 0.0
(01) Max grad: 0.0
(02) Max grad: 0.0
(03) Max grad: 0.0
(04) Max grad: 0.0
(05) Max grad: 0.0
(06) Max grad: 0.0
(07) Max grad: 0.0
(08) Max grad: 0.0
(09) Max grad: 0.0
(10) Max grad: 0.0
(11) Max grad: 0.0
(12) Max grad: 0.0
(13) Max grad: 0.0
(14) Max grad: 2.55876088142395
(15) Max grad: 0.0
(16) Max grad: 0.0
(17) Max grad: 0.0
(18) Max grad: 0.0
(19) Max grad: 0.0
```

## Expected behavior

All gradients should be zero.

## Environment

```
PyTorch version: 1.7.0
Is debug build: True
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Fedora 33 (Thirty Three) (x86_64)
GCC version: (GCC) 10.2.1 20201125 (Red Hat 10.2.1-9)
Clang version: 11.0.0 (Fedora 11.0.0-2.fc33)
CMake version: version 3.18.4

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce GTX 1050 Ti
Nvidia driver version: 460.32.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.2
[pip3] torch==1.7.0
[pip3] torchvision==0.8.1
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               10.2.89              hfd86e86_1  
[conda] libblas                   3.8.0                    21_mkl    conda-forge
[conda] libcblas                  3.8.0                    21_mkl    conda-forge
[conda] liblapack                 3.8.0                    21_mkl    conda-forge
[conda] liblapacke                3.8.0                    21_mkl    conda-forge
[conda] mkl                       2020.2                      256  
[conda] mkl-service               2.3.0            py38he904b0f_0  
[conda] mkl_fft                   1.2.0            py38h23d657b_0  
[conda] mkl_random                1.1.1            py38h0573a6f_0  
[conda] numpy                     1.19.2           py38h54aff64_0  
[conda] numpy-base                1.19.2           py38hfa32c7d_0  
[conda] pytorch                   1.7.0           py3.8_cuda10.2.89_cudnn7.6.5_0    pytorch
[conda] torchvision               0.8.1                py38_cu102    pytorch
```

## Additional context

- I've tested this on different GPU's. All of them gave errors, but not always the same error and not always for the same seed.
- On a same GPU, the same seed might not always give an error depending on the used seed `range` in the code sample.

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @ngimel

**Comment by mruberry** [comment]
We would accept a fix for this issue.

**Comment by CedricPicron** [comment]
@jjsjann123 Any idea what might be causing this issue?

**Comment by ngimel** [comment]
The following simplified script 
```
import torch
import torch.nn.functional as F


in_shape = (45,18) #torch.randint(low=10, high=50, size=(2,)).tolist()
in_map = torch.ones(2, 256, *in_shape, device='cuda', requires_grad=True)

out_shape = torch.randint(low=10, high=50, size=(2,)).tolist()
out_shape = (14,24)
inter_map = F.interpolate(in_map, size=tuple(out_shape), mode='nearest')
output_map = torch.mul(inter_map, 0.0)
output_map.sum().backward()

max_grad = in_map.grad.abs().max().item()
print(f"Max grad: {max_grad}")
```
reports cuda-memcheck error when disabling caching allocator:
```
(pt) ngimel@devfair0202:~/pytorch$ PYTORCH_NO_CUDA_MEMORY_CACHING=1 cuda-memcheck python ~/playground/interpolate_grad.py 
========= CUDA-MEMCHECK
========= Invalid __global__ read of size 4
=========     at 0x00000b90 in void at::native::_GLOBAL__N__52_tmpxft_001a17a9_00000000_7_UpSampleNearest2d_cpp1_ii_0758b255::upsample_nearest2d_backward_out_frame<float, float>(float const *, unsigned long, unsigned long, unsigned long, unsigned long, unsigned long, unsigned long, at::native::_GLOBAL__N__52_tmpxft_001a17a9_00000000_7_UpSampleNearest2d_cpp1_ii_0758b255::upsample_nearest2d_backward_out_frame<float, float>*, float, float)
=========     by thread (509,0,0) in block (404,0,0)
=========     Address 0x7f391fb50050 is out of bounds
=========     Saved host backtrace up to driver entry point at kernel launch time
=========     Host Frame:/lib/x86_64-linux-gnu/libcuda.so.1 (cuLaunchKernel + 0x34e) [0x2e46de]
=========     Host Frame:/public/apps/cuda/11.0/lib64/libcudart.so.11.0 [0xf62b]
=========     Host Frame:/public/apps/cuda/11.0/lib64/libcudart.so.11.0 (cudaLaunchKernel + 0x1c1) [0x4f5b1]
=========     Host Frame:/private/home/ngimel/pytorch/torch/lib/libtorch_cuda.so [0x28d5fce]
=========     Host Frame:/private/home/ngimel/pytorch/torch/lib/libtorch_cuda.so (_ZN2at6native32upsample_nearest2d_backward_cudaERKNS_6TensorEN3c108optionalINS4_8ArrayRefIlEEEES7_NS5_INS6_IdEEEE + 0xfa) [0x28d720a]
=========     Host Frame:/private/home/ngimel/pytorch/torch/lib/libtorch_cuda.so [0x32b155c]
=========     Host Frame:/private/home/ngimel/pytorch/torch/lib/libtorch_cuda.so [0x32b161c]
=========     Host Frame:/private/home/ngimel/pytorch/torch/lib/libtorch_cpu.so (_ZN2at27upsample_nearest2d_backwardERKNS_6TensorEN3c108optionalINS3_8ArrayRefIlEEEES6_NS4_INS5_IdEEEE + 0x20b) [0x1a17d6b]
=========     Host Frame:/private/home/ngimel/pytorch/torch/lib/libtorch_cpu.so [0x32dc034]
=========     Host Frame:/private/home/ngimel/pytorch/torch/lib/libtorch_cpu.so [0x32dc4ec]
=========     Host Frame:/private/home/ngimel/pytorch/torch/lib/libtorch_cpu.so (_ZN2at27upsample_nearest2d_backwardERKNS_6TensorEN3c108o
```

**Comment by jjsjann123** [comment]
Quick update.
1. I don't have the repro in ToT code reported;
2. I can verify the memory access issue in Natalia's example with cudamemcheck. This seems to be an oversight here: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/UpSample.cuh#L256, where we forgot to clamp it as we did with forward: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/UpSample.cuh#L246
index computation with floating point is fun :)
